doc/how-to-analyze-haskell-code-performance.rst

```diff
@@ -0,0 +1,97 @@
+How to analyze Haskell performance
+==================================
+
+When a Haskell application is slow or uses too much memory,
+Cabal and the
+`Haskell GHC compiler <https://downloads.haskell.org/ghc/latest/docs/users_guide/>`__
+can help you understand why.
+
+The main steps are to
+1. let GHC insert performance measuring code into your application,
+2. run the application to collect a performance report and
+3. visualize that report.
+
+The process of inserting performance measuring code and collecting performance information
+is called "profiling" and Cabal makes this is easy to do:
+
+Profiling CPU performance
+-------------------------
+
+First, build your application, e.g. ``my-app``, with profiling enabled:
+
+.. code-block:: console
+
+    $ cabal build --enable-profiling --profiling-detail=late exe:my-app
+
+Under the hood this will pass the corresponding
+`profiling compiler options <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#compiler-options-for-profiling>`__
+to GHC. With the additional Cabal flag ``--profiling-detail=late`` GHC is instructed to use
+`"late-cost-center" profiling <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#ghc-flag--fprof-late>`__
+and insert performance measuring code only after important optimisations
+have been applied to minimize the impact of profiling on performance.
+See the Cabal section on :ref:`profiling options <profiling-options>` for more details.
+
+Second, run the application with the necessary
+`runtime system (RTS) options <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html>`__
+to produce a profile report.
+
+.. code-block:: console
+
+    $ cabal list-bin exe:my-app
+    /path/to/my-app
+    $ /path/to/my-app +RTS -pj -RTS
+    <program runs and finishes>
+
+The report is written to a ``<app-name>.prof`` file, i.e. ``my-app.prof``, in the current directory.
+With the RTS option ``-pj`` the app produces a
+`"profile JSON" (pj) file report <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#rts-flag--pj>`__.
+Other report format options can be found in the
+`GHC format documentation. <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#time-and-allocation-profiling>`__.
+
+Finally, load the profiling report file into a visualizer to look for performance bottlenecks.
+One popular open-source
+`flame graph <https://www.brendangregg.com/flamegraphs.html>`__
+visualizer is `Speedscope <https://speedscope.app>`, which runs in the browser and comes with
+an example.
+
+Profiling your dependencies too
+-------------------------------
+
+The setup so far only profiles your main application, which is usually what you want.
+This happens by default, because Cabal command line options only apply to local packages
+and dependencies are usually not local.
+However, the bottlenecks may be in your dependencies so you would want to profile those too.
```

Comment:
```suggestion
However, the bottlenecks may be in your dependencies, so you would want to profile those too.
```

Reply:
Also, doesn't GHC require profiled builds of dependencies to go along with the profiled build of your application (see compiler ways)?

Reply:
> Also, doesn't GHC require profiled builds of dependencies to go along with the profiled build of your application

AFAIU, it requires the dependencies to be compiled with `-prof`, but you don't need to insert any SCCs there if you don't want to.

Reply:
FWIW, GHC manual is not very explicit about that, in https://downloads.haskell.org/ghc/9.10-latest/docs/users_guide/profiling.html it says

> Re-compile your program for profiling with the -prof option, and probably one of the options for adding automatic annotations: -fprof-late is the recommended option.

That implies that you need to recompiled everything; but I cannot find any explicit mention that mixing (i.e. linking) `-prof` and non-`-prof` objects would not work. 

It wont, you may get a compile time error like

```
ghc -c Bar.hs      

Bar.hs:3:1: error:
    Bad interface file: ./Foo.hi
        mismatched interface file profile tag (wanted "", got "p")
  |
3 | import Foo
  | ^^
```

but again, it's not mentioned in the documentation. (Though probably only cabal developers would be interested in such sort of information, I think it's good to exist).

Reply:
I think the closest it gets is the manual quote I put later (https://github.com/haskell/cabal/pull/10539#discussion_r1855251873). But that could be taken to mean that profiling would be ineffective, as opposed to causing compilation failure due to mismatched ways.

Reply:
> but I cannot find any explicit mention that mixing (i.e. linking) -prof and non--prof objects would not work.

I guess it's true that technically the way it's phrased does not specify this. If someone is willing to open a ticket suggesting a more accurate phrasing on the ghc issue tracker I can take care of the actual MR. Just assign me to the ticket if you do.

Reply:
https://gitlab.haskell.org/ghc/ghc/-/issues/25518 and I've assigned you (assuming you're AndreasK on GitLab).